base/essentials.jl

```diff
@@ -775,6 +775,9 @@ inferencebarrier(@nospecialize(x)) = Ref{Any}(x)[]
 
 Determine whether a collection is empty (has no elements).
 
+This operation is guaranteed not to consume elements even for stateful
+iterators (which should define [`Base.isdone`](@ref) to that end).
```

Comment:
What about this then?
```suggestion
!!! warning

    `isempty(itr)` may consume the next element of a stateful iterator `itr`
    unless an appropriate [`Base.isdone(itr)`](@ref) method is defined.
    Use of `isempty` should therefore be avoided when writing generic
    code which should support any iterator type.
```

Reply:
LGTM :+1:

Reply:
How about "unless an appropriate Base.isdone or isempty method is defined"?

Or "The fallback isempty(itr) method may consume..."?

As you've written it is also okay.

Reply:
OK, let's go with this then:
```suggestion
!!! warning

    `isempty(itr)` may consume the next element of a stateful iterator `itr`
    unless an appropriate [`Base.isdone(itr)`](@ref) or `isempty` method is defined.
    Use of `isempty` should therefore be avoided when writing generic
    code which should support any iterator type.
```